### PR TITLE
Chore: Remove lodash

### PR DESCRIPTION
**Pull request by stephenwade**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain: _Reduce dependencies_

#### What changes did you make? (Give an overview)

Addresses part of #14098. Removes lodash entirely from the project, replacing all uses with native code, custom code, and a handful of other tiny modules.

<details>
<summary>node_modules size comparison</summary>

```diff
~/git/eslint-test-master$ echo '{ "dependencies": { "eslint": "eslint/eslint" } }' > package.json
~/git/eslint-test-master$ npm install >/dev/null
~/git/eslint-test-master$ du -sh node_modules
22M	node_modules

~/git/eslint-test-stephenwade$ echo '{ "dependencies": { "eslint": "stephenwade/eslint#remove-lodash-3" } }' > package.json
~/git/eslint-test-stephenwade$ npm install >/dev/null
~/git/eslint-test-stephenwade$ du -sh node_modules
17M	node_modules

~/git$ diff -U0 <(cd eslint-test-master; du -hd1 node_modules) <(cd eslint-test-stephenwade; du -hd1 node_modules) | grep "^[-+]\d"
-4.9M	node_modules/lodash
-16K	node_modules/escape-string-regexp
+20K	node_modules/escape-string-regexp
+28K	node_modules/deep-extend
-240K	node_modules/@babel
+256K	node_modules/@babel
-22M	node_modules
+17M	node_modules
```
</details>

#### Is there anything you'd like reviewers to focus on?

There is one fairly major change in here. I replaced lodash.template by creating a JS file in place of each template. Let me know if I should take a different approach with that or if I should split it out into a separate PR.

### CODE REVIEW

**Comment by stephenwade** [comment]
I'll take a look at the CI failure tomorrow. Looks like I missed a .flat call

**Review by bcomnes — APPROVED** [review]

File: package.json
Comment (g-plane) [inline package.json:53]: Recommend: [rfdc](https://github.com/davidmarkclements/rfdc). `clone` hasn't been updated about three years.

File: tools/eslint-fuzzer.js
Comment (fregante) [inline tools/eslint-fuzzer.js:13]: I don't think this is a useful dependency. Now it can be replicated with:

```js
Object.fromPairs(Object.entries(obj).map(([value, key]) => cb(value, key, obj))))
```

Or likely something easier to read and more specific to the situation

File: package.json
Comment (fregante) [inline package.json:89]: This change should be dropped since `table` will be dropped by https://github.com/eslint/eslint/pull/14316

File: tests/lib/rules/no-invalid-this.js
Comment (fregante) [inline tests/lib/rules/no-invalid-this.js:83]:
```suggestion
        delete thisPattern.valid;
        delete thisPattern.invalid;
        return thisPattern;
```

File: lib/rule-tester/rule-tester.js
Comment (fregante) [inline lib/rule-tester/rule-tester.js:484]: Not as straightforward, but lets us skip a barely-used dependency
```suggestion
                const itemConfig = {...item};
                for (const parameter of RuleTesterParameters) {
                	delete[parameter];
                }
```

File: package.json
Comment (fregante) [inline package.json:85]: I think `deepExtend({}, obj)` is equivalent to `clone(obj)`. Can this dependency be dropped too?

File: package.json
Comment (mdjermanovic) [reply package.json:89]: It's fine to have this change included since this PR can be merged for a minor or patch release, while https://github.com/eslint/eslint/pull/14316 has to wait for the next major.

File: tools/eslint-fuzzer.js
Comment (stephenwade) [reply tools/eslint-fuzzer.js:13]: I would prefer to use the dependency. I think

```js
mapValues(ruleConfigs, sample)
```

is way more readable than

```js
Object.fromEntries(Object.entries(ruleConfigs).map(([value, key]) => [value, sample(key)]))
```

In my opinion, the addition of a tiny (16K) module is worth the readability here.

Comment (fregante) [reply tools/eslint-fuzzer.js:13]: Why use a “tiny 16k dependency” when you can use a 1-liner in some other file? It makes zero sense for a PR that attempts to reduce dependencies. 

Like I said, that’s the exact equivalent but you can find better solutions for the 2 places where it’s used.

Comment (stephenwade) [reply tools/eslint-fuzzer.js:13]: I hadn't thought of creating a function. That's a good idea. I'll do that and remove the dependency.

Comment (stephenwade) [reply tools/eslint-fuzzer.js]: Object.fromEntries is not available until Node 12. I have added a TODO comment to replace this dependency with a helper function when we drop support for Node 10.

Resolved in a28cf2db1.

File: lib/rules/comma-dangle.js
Comment (aladdin-add) [inline lib/rules/comma-dangle.js:155]: there is a stage-3 proposal: https://github.com/tc39/proposal-relative-indexing-method
so it will be written to `array.at(-1)` someday. 😄

File: tests/lib/rules/no-invalid-this.js
Comment (stephenwade) [reply tests/lib/rules/no-invalid-this.js]: Resolved in ada00d52f.

File: lib/rule-tester/rule-tester.js
Comment (stephenwade) [reply lib/rule-tester/rule-tester.js]: Resolved in ada00d52f.

File: lib/linter/apply-disable-directives.js
Comment (stephenwade) [reply lib/linter/apply-disable-directives.js:127]: I am hoping this PR can be merged before the next major release. If Node 10 support is dropped before this is merged, I will update this PR at that time.

File: package.json
Comment (stephenwade) [reply package.json]: Resolved in f9ac034a2.

File: package.json
Comment (stephenwade) [reply package.json]: Resolved this in 7dff2566a by adding `lib/shared/string-utils`.

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks!

**Review by mdjermanovic — CHANGES_REQUESTED** [review]
@stephenwade thanks for all this work! I left a couple of suggestions, didn't manage to review all details yet.

File: lib/shared/deprecation-warnings.js
Comment (stephenwade) [reply lib/shared/deprecation-warnings.js]: Thanks for catching that! I've removed the fast-memoize dependency in 6232136f3.

File: tests/lib/shared/string-utils.js
Comment (stephenwade) [reply tests/lib/shared/string-utils.js:41]: Done in e46f4716c.

File: lib/rules/object-curly-newline.js
Comment (mdjermanovic) [inline lib/rules/object-curly-newline.js:78]: In the context of this rule, we just need to check if the given value is an object. Can we do something like this, to not rely on `"[object Object]"`:

```js
/**
 * Check if a value is an object.
 * @param {any} value The value to check
 * @returns {boolean} `true` if the value is an object, otherwise `false`
 */
function isObject(value) {
    return typeof value === "object" && value !== null;
}
```

File: lib/linter/node-event-generator.js
Comment (mdjermanovic) [inline lib/linter/node-event-generator.js:197]: Can we use `Map` instead of object here? Selectors can be user-defined (e.g., in `no-restricted-syntax` rule), so it's better to avoid usual issues with object-maps.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/indent.js
Comment (stephenwade) [reply lib/rules/indent.js]: Resolved in c20cce3eb.

File: lib/rules/object-curly-newline.js
Comment (stephenwade) [reply lib/rules/object-curly-newline.js]: Resolved in a1bd47baa.

File: lib/linter/node-event-generator.js
Comment (stephenwade) [reply lib/linter/node-event-generator.js]: Resolved in a1bd47baa.

File: package.json
Comment (mdjermanovic) [inline package.json:55]: There are some differences between `deep-extend` and `lodash.merge`/`lodash.cloneDeep`:

```js
const lodash = require("lodash");
const extend = require("deep-extend");

const test1 = [{ foo: 5 }, { foo: undefined }];

console.log(lodash.merge({}, ...test1)); // { foo: 5 }
console.log(extend({}, ...test1)); // { foo: undefined }

const test2 = [{ foo: [1, 2, 3] }, { foo: [4] }];

console.log(lodash.merge({}, ...test2)); // { foo: [4, 2, 3] }
console.log(extend({}, ...test2)); // { foo: [4] }

const test3 = { foo: {} };
test3.foo.bar = test3; // circular

lodash.cloneDeep(test3); // ok
extend({}, test3); // RangeError: Maximum call stack size exceeded

```

We should estimate the impact of the differences, or use some other dependency.

The latest `0.6.0` works the same as `0.5.1` on these examples. Also, was there a specific reason to require `^0.5.1` instead of `^0.6.0`?

Comment (stephenwade) [reply package.json]: > Also, was there a specific reason to require `^0.5.1` instead of `^0.6.0`?

Weird, that's what happens when I run `npm i -s deep-extend`. I've bumped this up to `^0.6.0` in b3d6d9a7b.

Comment (stephenwade) [reply package.json]: #

Comment (stephenwade) [reply package.json]: I tested 8 different Node modules for deep merging. None of them had quite the same behavior as `lodash`.

<details>
<summary>Differences in behavior between various Node modules for deep merging</summary>

## `lodash.merge`

```jsx
const f = require("lodash").merge;

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: 5 }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [4, 2, 3] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// { foo: <ref *1> { bar: { foo: [Circular *1] } } }
```

## `deep-extend`

```jsx
const f = require("deep-extend");

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: undefined }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// Uncaught RangeError: Maximum call stack size exceeded
```

## `deepmerge`, `merge-deep`

```jsx
const deepmerge = require("deepmerge");
const f = (...params) => deepmerge.all(params);
// or
const f = require("merge-deep");

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: undefined }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 1, 2, 3, 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// Uncaught RangeError: Maximum call stack size exceeded
```

## `@common-utilities/merge-objects`

```jsx
const { mergeObjects } = require("@common-utilities/merge-objects");
const f = (...params) => params.reduce(mergeObjects);

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: 5 }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 1, 2, 3, 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// { foo: <ref *1> { bar: { foo: [Circular *1] } } }
```

## `assign-deep`, `merge`

```jsx
const f = require("assign-deep");
// or
const f = require("merge").recursive;

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: undefined }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// { foo: <ref *1> { bar: { foo: [Circular *1] } } }
```

## `deep-assign-lite`

```jsx
const f = require("deep-assign-lite");

// test undefined values
f({}, { foo: 5 }, { foo: undefined });
// { foo: 5 }

// test arrays
f({}, { foo: [1, 2, 3] }, { foo: [4] });
// { foo: [ 4 ] }

// test circular referencces
const obj = { foo: {} };
obj.foo.bar = obj;
f({}, obj);
// Uncaught RangeError: Maximum call stack size exceeded
```

---
</details>

I'll look into where this function is used to see what the differences would mean for actual usage.

Comment (stephenwade) [reply package.json]: Below I've evaluated the behavior changes of merging undefined values, merging arrays, and circular references as they apply to each use of `extend`.

<details>
<summary>lib/linter/linter.js</summary>

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/linter/linter.js#L530-L532

- Undefined values could be an issue if a plugin defines an environment with an `undefined` value somewhere.
- Array merging and circular references are not an issue since the [`ParserOptions` type](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/shared/types.js#L14-L26) does not allow arrays or objects as values.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/linter/linter.js#L533

- Undefined values could be an issue if the user provides a `parserOptions` object with an `undefined` value somewhere.
- Array merging and circular references are not an issue since the [`ParserOptions` type](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/shared/types.js#L14-L26) does not allow arrays or objects as values.

---
</details>

<details>
<summary>lib/rule-tester/rule-tester.js</summary>

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L328-L333

- Undefined values, array merging, and circular references are not issues. I looked at all the uses of the `RuleTester` constructor and none of them use `undefined`, arrays, or circular references.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L372

- Undefined values and array merging are not an issue since one of the parameters is an empty object.
- Circular references are not an issue since `testerDefaultConfig` is [defined in this file](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L112) and does not contain circular references.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L468

- Undefined values, array merging, and circular references are not issues. `testerConfig` is defined in the `RuleTester` constructor, and none of the uses of the `RuleTester` constructor use `undefined`, arrays, or circular references.

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/lib/rule-tester/rule-tester.js#L490-L493

- I can't look through all tests to check the config on each one, but since the tests still pass I don't think any of them use `undefined` or circular references.
- Array merging is not an issue since `config` is a clone of `testerConfig` which is defined in the `RuleTester` constructor, and none of the uses of the `RuleTester` constructor use arrays.

---
</details>

<details>
<summary>tests/lib/rules/no-invalid-this.js</summary>

https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/tests/lib/rules/no-invalid-this.js#L73

- Undefined values and array merging are not an issue since one of the parameters is an empty object.
- Circular references are not an issue since each `pattern` is [defined in this file](https://github.com/eslint/eslint/blob/b3d6d9a7b2108e5de4fccbde248f0b4fae00254b/tests/lib/rules/no-invalid-this.js#L103) and no pattern contains circular references.

---
</details>

Summary: I believe that the differences in array merging and circular references will not affect end users. The differences in merging `undefined` values might affect users of `parserOptions` and custom environments.

There are 2 replacements I [tested](https://github.com/eslint/eslint/pull/14287#discussion_r628662466) that ignore `undefined` values, and neither are very popular by download count. Of the 2 options, I think [@common-utilities/merge-objects](https://www.npmjs.com/package/@common-utilities/merge-objects) is the better option, but I would understand not wanting to use either one based on their download counts.

In my view, the options are:
1. Switch to a not-popular package for merging objects so the `undefined` behavior doesn't change.
2. Accept that the behavior of `undefined` will change and wait until ESLint v8 to merge this PR (since it would be a breaking change).

/cc @mdjermanovic What are your thoughts?

Comment (mdjermanovic) [reply package.json]: I'd guess that none of the changes from `lodash.merge/lodash.cloneDeep` to `deep-extend` would cause a significant impact, but to be safe, maybe we can use [lodash.merge package](https://www.npmjs.com/package/lodash.merge) in this PR (which could be merged for ESLint v7), and replace it with `deep-extend` in ESLint v8.0.0.

@stephenwade @aladdin-add @eslint/eslint-tsc thoughts?

Comment (nzakas) [reply package.json]: I think we can probably just use lodash.merge. I’m not sure there’s a value in doing another switch after that because I think all of this will end up handled inside FlatConfigArray, so we will be able to remove any other merge-related dependencies when eslintrc is purged.

Comment (stephenwade) [reply package.json]: I think it would make sense to use `lodash.merge` for now, but use of the per-method packages is [discouraged by Lodash](https://lodash.com/per-method-packages) so I think we should switch away from them in ESLint v8.0.0.

I saw that a [`deepMerge` function](https://github.com/eslint/eslint/pull/14321/files#diff-93aef33006c32545a3aa83cf3dfedd4959e55327e14abf62744aa845c5d789d8R54-R60) is going to be included in the new config system (#14321). Could we extract that into `lib/shared` and use it for v8 instead of `lodash.merge` or `deep-extend`?

cc: @nzakas @mdjermanovic

Comment (nzakas) [reply package.json]: We could. I haven’t verified that it works exactly the same as lodash.merge, though. So we would need to do some further testing. If you’re up for it, I’d say go for it.

File: lib/source-code/token-store/utils.js
Comment (mdjermanovic) [inline lib/source-code/token-store/utils.js:36]: Similar to the previous comment, `lodash.sortedIndexBy` most likely uses a binary search (and jsdoc description for this `function search` says "Binary-searches..."), while `findIndex` performs a linear search.

I think `utils.search` is used only on arrays of comments, which typically shouldn't be large arrays, so the difference _might_ be small.

@eslint/eslint-tsc thoughts about this, should we look for a binary search library?

Comment (mysticatea) [reply lib/source-code/token-store/utils.js:36]: I also think the difference might be small.

By the way, `getStartLocation({ range: [location] })` is `{ range: [location] }.range[0]`, so it does nothing.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/source-code/source-code.js
Comment (stephenwade) [reply lib/source-code/source-code.js]: Suggestion committed in cba69f23c.

File: lib/source-code/token-store/utils.js
Comment (stephenwade) [reply lib/source-code/token-store/utils.js]: Useless `val` declaration removed in 00ac18335.

File: lib/source-code/source-code.js
Comment (stephenwade) [reply lib/source-code/source-code.js]: findIndex is slower, but not exponentially so, and both operations can run millions of times per second.

```
$ node benchmark-getLocFromIndex.js
sortedLastIndex x 10,168,021 ops/sec ±1.34% (90 runs sampled)
Array#findIndex x 2,795,631 ops/sec ±1.41% (89 runs sampled)
```

Benchmark code: https://github.com/stephenwade/my-eslint-benchmark

File: lib/source-code/token-store/utils.js
Comment (stephenwade) [reply lib/source-code/token-store/utils.js]: findIndex is slower, but not exponentially so, and both operations can run millions of times per second.

findIndex is slower, but not exponentially so, and both operations can run millions of times per second.

```
$ node benchmark-search.js
sortedIndexBy x 7,188,306 ops/sec ±0.84% (89 runs sampled)
Array#findIndex x 2,764,162 ops/sec ±1.65% (84 runs sampled)
```

Benchmark code: https://github.com/stephenwade/my-eslint-benchmark

Comment (stephenwade) [reply lib/source-code/token-store/utils.js]: I used [hyperfine](https://github.com/sharkdp/hyperfine) to benchmark the test suite on this branch and the master branch. This branch ran slightly faster on my computer, although the results were very close.

Screenshot of benchmark results running the npm test command. The remove-lodash-3 branch runs in about 172 seconds and the master branch runs in about 174 seconds.

I believe that the speed difference in this method will not be noticeable and that we should stick with findIndex.

File: lib/source-code/source-code.js
Comment (stephenwade) [reply lib/source-code/source-code.js]: See https://github.com/eslint/eslint/pull/14287/commits/02510b2c35222d12741ac469753156e86a41f844#r636050582 for project-wide benchmarking results. I believe the speed difference in this function will not be noticeable.

File: lib/cli-engine/formatters/html-template-message.js
Comment (stephenwade) [reply lib/cli-engine/formatters/html-template-message.js:21]: Resolved in e4d94ce7c.

File: lib/cli-engine/formatters/html.js
Comment (stephenwade) [reply lib/cli-engine/formatters/html.js]: Resolved in e4d94ce7c.

Comment (stephenwade) [reply lib/cli-engine/formatters/html.js]: I had copied the list of characters from [doT](https://github.com/olado/doT/blob/031d3bb7520eed6b93886df2b650b7fce12a7007/doT.js#L33), but I don't think the `/` is necessary. I've removed it.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks for the great work!

**Comment by mdjermanovic** [comment]
Thanks for contributing!

**Pull request merged**
